src/browser/register.ts

```diff
@@ -1,18 +1,38 @@
-import { getOptions, normalize } from "../common/util"
-
-const options = getOptions()
+import { logger } from "@coder/logger"
+import { getOptions, Options, normalize } from "../common/util"
 
 import "./pages/error.css"
 import "./pages/global.css"
 import "./pages/login.css"
 
-if ("serviceWorker" in navigator) {
-  const path = normalize(`${options.csStaticBase}/dist/serviceWorker.js`)
-  navigator.serviceWorker
-    .register(path, {
+export async function registerServiceWorker(navigator: Navigator, path: string, options: Options): Promise<void> {
+  try {
+    await navigator.serviceWorker.register(path, {
       scope: (options.base ?? "") + "/",
     })
-    .then(() => {
-      console.log("[Service Worker] registered")
-    })
+    console.log("[Service Worker] registered")
+  } catch (error) {
+    logger.error(`[Service Worker] failed to register: ${error.message}`)
+  }
+}
+
+interface HandleServiceWorkerRegistration {
+  getOptions: () => Options
+  normalize: (url: string, keepTrailing?: boolean) => string
+  registerServiceWorker: (navigator: Navigator, path: string, options: Options) => Promise<void>
 }
+
+export function handleServiceWorkerRegistration({
+  getOptions,
+  normalize,
+  registerServiceWorker,
+}: HandleServiceWorkerRegistration): void {
+  if (typeof navigator !== "undefined" && "serviceWorker" in navigator) {
+    const options = getOptions()
+    const path = normalize(`${options.csStaticBase}/dist/serviceWorker.js`)
+    registerServiceWorker(navigator, path, options)
+  }
+}
+
+// Written this way so that it's easier to test
+handleServiceWorkerRegistration({ getOptions, normalize, registerServiceWorker })
```

Comment:
I don't _love_ this function signature (taking in an object with these three keys), but it makes it super easy to test 🤷‍♂️

Reply:
It does feel a little strange to me as well. Could we export a single `registerServiceWorker` that takes no parameters?

Then we polyfill `navigator`, `document.getElementById`, and `getAttribute` (the last two are used in `getOptions` so we can use that to test different base values). That would make the test most closely mirror how the code runs in the browser, I think.

Reply:
We could also do no export and `require` the module each time and clear the module cache after each test.